src/control/cmd/daos/snapshot.go

```diff
@@ -129,11 +133,71 @@ func (cmd *containerSnapListCmd) Execute(args []string) error {
 	}
 	defer cleanup()
 
-	rc := C.cont_list_snaps_hdlr(ap)
+	var anchor C.daos_anchor_t
+	var snapCount C.int
+
+	rc := C.daos_cont_list_snap(ap.cont, &snapCount, nil, nil, &anchor, nil)
 	if err := daosError(rc); err != nil {
-		return errors.Wrapf(err,
-			"failed to list snapshots for container %s",
-			cmd.contUUID)
+		return errors.Wrapf(err, "failed to count snapshots for container %s", cmd.ContainerID())
+	}
+
+	if !C.daos_anchor_is_eof(&anchor) {
+		return errors.New("too many snapshots returned")
+	}
+
+	epochs := make([]C.daos_epoch_t, snapCount)
+	names := make([]*C.char, snapCount)
+	defer func() {
+		for _, name := range names {
+			freeDaosStr(name)
+		}
+	}()
+	expectedSnaps := snapCount
+	C.memset(unsafe.Pointer(&anchor), 0, C.sizeof_daos_anchor_t)
+	if snapCount > 0 {
+		rc = C.daos_cont_list_snap(ap.cont, &snapCount, &epochs[0], &names[0], &anchor, nil)
+		if err := daosError(rc); err != nil {
+			return errors.Wrapf(err, "failed to list snapshots for container %s", cmd.ContainerID())
+		}
+	}
+
+	if expectedSnaps < snapCount {
+		return errors.New("snapshot list has been truncated (size changed)")
```

Comment:
Nit, the message is a little bit confusing for me, as the list of snapshots was not truncated, but expanded with new snaps. However, I agree that some snapshot information are missing in epochs and names structure.